File: leetcode3/이진희/3954. Sum of Compatible Numbers in Range I.java
/*

1. 아이디어 : 수식을 만족하는 x의 값 구하기

2. 시간복잡도 : O(200) // n=100 k=100

3. 자료구조/알고리즘 : 수학

 */

class Solution {
    public int sumOfGoodIntegers(int n, int k) {
        // x는 양수 
        // k+n >= x >= n-k
        int ans = 0;
        int start = n-k>0 ? n-k : 1;
        int end = n+k;
        
        for(int i = start; i<= end; i++) {
            if((n&i) == 0) ans+=i;
        }

        return ans;
    }
}

File: leetcode3/최원준/3954. Sum of Compatible Numbers in Range I.py
#

'''
1. 아이디어 :
-

2. 시간복잡도 :
    O(2k)

3. 자료구조/알고리즘 :
-

'''
class Solution:
    def sumOfGoodIntegers(self, n: int, k: int) -> int:
        ans = 0
        x = max(1, n-k)
        while abs(n-x) <= k:
            if n&x == 0:
                ans += x
            x+=1
        return ans
